decision-records/2024-08-29--dashboard-uid-collision.md

```diff
@@ -0,0 +1,134 @@
+
+#### (b) New dashboard created by cloning an existing dashboard, neglecting to modify the UID
+
+A common way to create a new dashboard is to clone and modify dashboard (local or from the grafana marketplace).
+If the UID isn't modified (need to remeber to do it manually), then we'll end up with a UID collision (but the revision would be auto-incremented by the grafana UI, if saved and exported using grafana).
+
+#### (c) Unrelated dashboards that coincidentally happen to have the same UID
+
+This is likely to happen when charm authors manually modify UIDs, assuming their (naive :) modification produced a unique ID (in the ecosystem!).
+
+
+## Requirements
+1. Dashboards must not have the same *uid* within installation.
+2. Dashboard URL must be stable across dashboard updates (e.g. charm upgrades). Do not generate the UID from hashing
+                                                                                                                                                                                                                                                                                                                                                                                                                             dashboard's entire contents, because it would break links to the dashboard.[^DbURL]
+3. UID length must be <= 40 (grafana limit).
+4. Should be able to easily tell from which charm each dashboard came.
+
+## Decision
+(2/P) + (4/R) + (5/P) + (6).
+
+## Alternatives considered
+Deduplication can be addressed on the provider (charm with dashboards) or the requrier side (grafana charm).
+In the following subsections, `P` and `R` stand for provider and requirer.
+
+### (1/P) Use a prescribed naming pattern for UIDs at dev time
+We can update the [best practices doc](https://discourse.charmhub.io/t/grafana-k8s-docs-how-to-create-a-great-charmed-dashboard/14188#use-an-effective-naming-scheme-22),
+and it's up to charm authors to be disciplined about adhering to it. Charm lib will raise if the pattern doesn't match the expected ("test-time linter").
+
+A good option for UIDs could be `f"{charm_name}-{codename}"`, where `codename` is e.g. an abbreviate dashboard title that the charm author needs to make sure is unique within the charm.
+
+Benefits:
+- UID is human-readable and sensible.
+
+Disadvantages:
+- Coming up with a codename is toilsome.
+- To adhere to this new standard, would need to modify all existing dashboards.
+
+### (2/P) Automatically overwrite UIDs with a prescribed naming pattern at deploy/relate time, on the provider side
+SHA1 is of length 40 (`len(hashlib.sha1(b"whatever").hexdigest()) == 40`), so calculate the dashboards UID from hashing (charm_name + relative_path of the dashboard file).
+
+Benefits:
+- The combination of charm_name and relative_path is almost guaranteed to be unique across the ecosystem. Exceptions:
+  - Problem (a)
```

Comment:
Is Problem (a) referring to: `(a) Two charms on different revisions`?

Reply:
Yes. Clarified.